6223/custom pages guide improvements (#6264)
keystonejs/keystone: docs/pages/docs/guides/custom-admin-ui-pages.mdx

# Custom Admin UI Pages

## Getting Started

In this guide we'll show you how to add custom pages to the Keystone Admin UI.
As the Admin UI is built on top of [Next.js](https://nextjs.org/docs/basic-features/pages), it exposes the same pages directory for adding custom pages.

> 
> !> **Note:** This feature will improve over time. It has been released ahead of time to unblock developers. We are working on improving support for other styling frameworks besides Emotion and will be making it easier by exporting more Admin UI components in the near future.

Reviewer: ```suggestion
```

I don't think this is helpful any more.

The `PageContainer` component takes a `header` prop, which is expected to be a `ReactElement`.
This `header` prop is rendered out as the page title at the top of the page.

```tsx
// admin/pages/custom-page.tsx
import { PageContainer } from '@keystone-next/keystone/admin-ui/components';
export default function CustomPage () {
    return (
        <PageContainer header="Custom Page">
            <h1>This is a custom Admin UI Page</h1>
            <p>It can be accessed via the route <a href="/custom-page">/custom-page</a></p>
        </PageContainer>
    )
}
```

With the above snippet, our custom page looks a lot more like the other pages in the Admin UI.
![example of the custom Admin UI page with the PageContainer component](/assets/guides/custom-admin-ui-pages/custom-page-w-page-container.png)

There's still a problem though, the header doesn't look right. If we compare the header of our custom page with the header for the Dashboard, there's quite a bit of difference in the styling and font-weight.
![example of the Dashboard header element](/assets/guides/custom-admin-ui-pages/header-prop.png)

> Keystone pages leverage the `Heading` component from the `@keystone-ui/core` package to style the header, so lets use this to give our header the same styling.

Reviewer: ```suggestion
Keystone pages leverage the `Heading` component from the `@keystone-ui/core` package to style the header, so let's use this to give our header the same styling.
```

## Custom route in Admin UI Navigation

Yes, our custom page is looking pretty great, and much more like an Admin UI page, but it's not visible as a navigation item.
We can fix this by adding a custom Navigation component with a route pointing to our custom page.

First add the following files to the `/admin` directory in the root of your Keystone project.

```tsx
// admin/config.ts
import type { AdminConfig } from '@keystone-next/types';
import { CustomNavigation } from './components/CustomNavigation';
export const components: AdminConfig['components']= {
    Navigation: CustomNavigation
};
```

```tsx
// admin/components/CustomNavigation.tsx
import { NavigationContainer, ListNavItems, NavItem } from '@keystone-next/keystone/admin-ui/components';
import type { NavigationProps } from '@keystone-next/keystone/admin-ui/components';
export function CustomNavigation({ lists, authenticatedItem }: NavigationProps) {
    return (
        <NavigationContainer authenticatedItem={authenticatedItem}>
            <NavItem href="/">Dashboard</NavItem>
            <ListNavItems lists={lists} />
        </NavigationContainer>
    )
}
```

> !> You will need to restart your Keystone system after adding `admin/config.ts` for the custom Navigation component to be loaded.

Reviewer: This isn't rendering correctly.

> !> We've skipped over the specifics for this guide. If you're interested in more details on creating a custom Navigation component check out the [Custom Admin UI Navigation](/docs/guides/custom-admin-ui-navigation) guide.

Reviewer: ```suggestion
!> If you're interested in more details on creating a custom Navigation component check out the [Custom Admin UI Navigation](/docs/guides/custom-admin-ui-navigation) guide.
```

Lastly we'll add our new route to the newly created `CustomNavigation` component.

```tsx
// admin/components/CustomNavigation.tsx
import { NavigationContainer, ListNavItems, NavItem } from '@keystone-next/keystone/admin-ui/components';
import type { NavigationProps } from '@keystone-next/keystone/admin-ui/components';
export function CustomNavigation({ lists, authenticatedItem }: NavigationProps) {
    return (
        <NavigationContainer authenticatedItem={authenticatedItem}>
            <NavItem href="/">Dashboard</NavItem>
            <ListNavItems lists={lists} />
            <NavItem href="/custom-page">Custom Page</NavItem>
        </NavigationContainer>
    )
}
```

> !> Under the hood Keystone's Admin UI is powered by Next.js, so the route to our custom page is the filename of our custom page component, in this case it's `/custom-page`

Reviewer: ```suggestion
!> Under the hood Keystone's Admin UI is powered by Next.js, so the route to our custom page is the filename of our custom page component. In this case it's `/custom-page`.
```

The default export of every file in this directory is expected to be a valid React Component rendered out as the contents of the route.

```tsx
// admin/pages/custom-page.tsx
export default function CustomPage () {
    return (
        <>
            <h1>This is a custom Admin UI Page</h1>
            <p>It can be accessed via the route <a href="/custom-page">/custom-page</a></p>
        </>
    )
}
```

x> **Not all Next.js exports are available:** Keystone **only** supports the page component as a default export in the pages directory. This means that unlike with Next, auxillary exports such as `getStaticProps` and `getServerProps` are not supported.

With this in place, we now have a nice simple custom Admin UI page at `http://localhost:3000/custom-page`.
![example of a simple custom-page in the Admin UI](/assets/guides/custom-admin-ui-pages/simple-custom-page.png)

## Adding Admin UI layout

> At the moment though, this page is pretty barebones. We want our page to look more like an Admin UI page.

Reviewer: ```suggestion
At the moment this page is pretty bare bones. We want our page to look more like an Admin UI page.
```

Keystone helps us do this via the `PageContainer` component exported from `@keystone-next/keystone/admin-ui/components`.

The `PageContainer` component takes a `header` prop, which is expected to be a `ReactElement`.
This `header` prop is rendered out as the page title at the top of the page.

```tsx
// admin/pages/custom-page.tsx
import { PageContainer } from '@keystone-next/keystone/admin-ui/components';
export default function CustomPage () {
    return (
        <PageContainer header="Custom Page">
            <h1>This is a custom Admin UI Page</h1>
            <p>It can be accessed via the route <a href="/custom-page">/custom-page</a></p>
        </PageContainer>
    )
}
```

With the above snippet, our custom page looks a lot more like the other pages in the Admin UI.
![example of the custom Admin UI page with the PageContainer component](/assets/guides/custom-admin-ui-pages/custom-page-w-page-container.png)

> There's still a problems though, the header doesn't look right. If we compare the header of our custom-page with the header for the Dashboard, there's quite a bit of difference in the styling and font-weight.

Reviewer: ```suggestion
There's still a problem though, the header doesn't look right. If we compare the header of our custom page with the header for the Dashboard, there's quite a bit of difference in the styling and font-weight.
```

## Custom route in Admin UI Navigation

Yes, our custom page is looking pretty great, and much more like an Admin UI page, but it's not visible as a navigation item.
We can fix this by adding a custom Navigation component with a route pointing to our custom page.

First add the following files to the `/admin` directory in the root of your Keystone project.

```tsx
// admin/config.ts
import type { AdminConfig } from '@keystone-next/types';
import { CustomNavigation } from './components/CustomNavigation';
export const components: AdminConfig['components']= {
    Navigation: CustomNavigation
};
```

```tsx
// admin/components/CustomNavigation.tsx
import { NavigationContainer, ListNavItems, NavItem } from '@keystone-next/keystone/admin-ui/components';
import type { NavigationProps } from '@keystone-next/keystone/admin-ui/components';
export function CustomNavigation({ lists, authenticatedItem }: NavigationProps) {
    return (
        <NavigationContainer authenticatedItem={authenticatedItem}>
            <NavItem href="/">Dashboard</NavItem>
            <ListNavItems lists={lists} />
        </NavigationContainer>
    )
}
```

> 

Reviewer: ```suggestion
!> You will need to restart your Keystone system after adding `admin/config.ts` for the custom Navigation component to be loaded.
```

> !> We've skipped over the specifics for this guide, if you're interested in more detail on creating a custom Navigation component check out the [Custom Admin UI Navigation](/docs/guides/custom-admin-ui-navigation) guide.

Reviewer: ```suggestion
!> We've skipped over the specifics for this guide. If you're interested in more details on creating a custom Navigation component check out the [Custom Admin UI Navigation](/docs/guides/custom-admin-ui-navigation) guide.
```

Lastly we'll add our new route to the newly created `CustomNavigation` component.

```tsx
// admin/components/CustomNavigation.tsx
import { NavigationContainer, ListNavItems, NavItem } from '@keystone-next/keystone/admin-ui/components';
import type { NavigationProps } from '@keystone-next/keystone/admin-ui/components';
export function CustomNavigation({ lists, authenticatedItem }: NavigationProps) {
    return (
        <NavigationContainer authenticatedItem={authenticatedItem}>
            <NavItem href="/">Dashboard</NavItem>
            <ListNavItems lists={lists} />
            <NavItem href="/custom-page">Custom Page</NavItem>
        </NavigationContainer>
    )
}
```

> !> Under the hood Keystone's Admin UI is powered by Next.js, so the route to our custom page is the filename of our custom page component, in this case its `/custom-page`

Reviewer: ```suggestion
!> Under the hood Keystone's Admin UI is powered by Next.js, so the route to our custom page is the filename of our custom page component, in this case it's `/custom-page`
```

With all that in place, our custom Admin UI page is now navigable from the Admin UI Navigation component, and we can access it from other pages in the Admin UI.
![completed custom Admin UI page](/assets/guides/custom-admin-ui-pages/custom-page-completed.png)

> !> This concludes the main part of the Custom Admin UI Pages guide
> 

Reviewer: ```suggestion
```

I think we can let people bounce from the document whenever they need without explicitly calling out checkpoints.

## Styling

> Of course there are other styling considerations when adding a custom page to the Admin UI that go beyond making it _look_ and _feel_ like an Admin UI page.

Reviewer: ```suggestion
There are other styling considerations when adding a custom page to the Admin UI that go beyond making it _look_ and _feel_ like an Admin UI page.
```

The snippet below uses the emotion `jsx` runtime exported from `@keystone-ui/core` to help add some basic allignment and layout styling to the contents of our Admin UI custom page.

```tsx
// admin/pages/custom-page.tsx
/** @jsxRuntime classic */
/** @jsx jsx */

import { jsx } from '@keystone-ui/core';
import { PageContainer } from '@keystone-next/keystone/admin-ui/components';
import { Heading } from '@keystone-ui/core';

export default function CustomPage () {
    return (
        <PageContainer
            header={(
                <Heading type="h3">
                    Custom Page
                </Heading>
            )}>
            <h1 css={{
                width: '100%',
                textAlign: 'center',
            }}>
                This is a custom Admin UI Page
            </h1>
            <p css={{ textAlign: 'center' }}>
                It can be accessed via the route <a href="/custom-page">/custom-page</a>
            </p>
        </PageContainer>
    )
}
```

> Of course this is purely a recommendation, if you would prefer to use another css-in-js or css solution for your custom component please feel free to! Although this may require additional configuration currently outside of the scope of this guide.

Reviewer: ```suggestion
Using `emotion` for styling is purely a recommendation, if you would prefer to use another css-in-js or css solution for your custom component please feel free to. This may require additional configuration currently outside of the scope of this guide.
```
